src/haz3lcore/dynamics/Constraint.re

```diff
@@ -119,15 +119,23 @@ let rec ctr_of_nth_variant = (num_variants, nth): (t => t) =>
     xi => InjR(xi |> ctr_of_nth_variant(num_variants - 1, nth - 1));
   };
 
-let of_ap = (ctx, ctr: option(Constructor.t), arg: t): t =>
+let of_ap = (ctx, mode, ctr: option(Constructor.t), arg: t): t =>
   switch (ctr) {
   | Some(name) =>
-    switch (Ctx.lookup_ctr(ctx, name)) {
-    | None => Falsity // TODO: review
-    | Some({num_variants, nth, _}) =>
-      arg |> ctr_of_nth_variant(num_variants, nth)
+    switch (mode) {
+    | Mode.Ana(ty) =>
+      switch (Typ.weak_head_normalize(ctx, ty)) {
+      | Sum(map) =>
+        let num_variants = ConstructorMap.cardinal(map);
+        switch (ConstructorMap.nth(name, map)) {
+        | Some(nth) => arg |> ctr_of_nth_variant(num_variants, nth)
+        | None => Falsity // TODO: review
+        };
+      | _ => Falsity // TODO: review
+      }
+    | _ => Falsity // TODO: review
```

Comment:
For `Mode.Syn` / `Mode.SynFun`, the constraint associated with an `Ap` pattern is not used.
An example in which an `Ap` pattern synthesizes a type is shown below:

![image](https://github.com/hazelgrove/hazel/assets/108375845/6a8a574a-2079-4e7d-a15e-1bf00039f433)

Reply:
we do still need to do exhaustiveness checking for patterns in lets and functions, so the type to be used here can be the synthesized type (which in the case of constructor applications I believe is just the singleton sum)